### ISSUE TITLE
getting incorrect unused_imports warning

### ISSUE THREAD

**Issue by Gozala** [rust-analyzer/rust-analyzer#3860]
Here is the an example code:

```rust
#![forbid(unsafe_code, future_incompatible, rust_2018_idioms)]
#![deny(missing_debug_implementations, nonstandard_style)]
#![recursion_limit = "512"]

#[allow(unused_imports)]
use proc_macro::TokenStream;
use quote::{quote, quote_spanned};
use syn::spanned::Spanned;

#[cfg(not(test))] // NOTE: exporting main breaks tests, we should file an issue.
#[proc_macro_attribute]
pub fn wait(_attr: TokenStream, item: TokenStream) -> TokenStream {
    let input = syn::parse_macro_input!(item as syn::ItemFn);

    let ret = &input.sig.output;
    let inputs = &input.sig.inputs;
    let name = &input.sig.ident;
    let body = &input.block;
    let attrs = &input.attrs;

    if input.sig.asyncness.is_none() {
        return TokenStream::from(quote_spanned! { input.span() =>
            compile_error!("the async keyword is missing from the function declaration"),
        });
    }

    let result = quote! {
        #(#attrs)*
        fn #name(#inputs) #ret {
          async_std::task::block_on(async {
            #body
          })
        }
    };

    result.into()
}
```

For which I am getting `unused imports` warnings even though they are all used:

<img width="922" alt="Screen Shot 2020-04-05 at 6 35 32 PM" src="https://user-images.githubusercontent.com/21236/78515848-52f23600-776c-11ea-9c00-7309db3a43cc.png">

Using version: 0.2.108

**Comment by matklad** [comment]
These warnings come from rustc, so we can't do anything about them on rustc-side.

This is a known problem: in test mode, the imports are indeed unused, because `#[cfg(not(test)]` code gets removed early in the compilation pipeline. It can be argued that this works as intended. I *think* there should be an issue on rust-lang/rust discussing this.
